### PR TITLE
Refactor `PublishingApiDocumentRepublishingWorker` and associated tests

### PR DESCRIPTION
**Pull request by yndajas**
## Tests

In working on #8993, we needed to extend the worker in order to support previewing which editions would be refreshed when calling `#perform`. We added some test coverage in #8998 in order to provide confidence that the refactor wouldn't break any existing functionality. However, when digging into the git history we found that most of these tests already existed before being removed in 56195d5. We agreed with the suggestion in that commit of making better use of mocking, so we've incorporated this into a wider rework of the tests which also aims to improve coverage (further) and structure

## Worker

Being relatively new to GDS/Publishing, we have a (relatively) low level of context on how everything works. With this in mind, we struggled a little to parse the worker class owing to the language and logic of both public and private methods. We've had a go at rewriting this into a more explicit, low context friendly format. We hope that it's now easier to understand the relationship between the various edition states and the actions that happen when calling `#perform`, and that this makes it a little easier to extend/modify the class in future

---

This was done as groundwork for #8993 ([Trello](https://trello.com/c/egFDeEYg))

---

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

### CODE REVIEW

File: app/workers/publishing_api_document_republishing_worker.rb [hunk 1]
```diff
@@ -12,111 +12,117 @@
...
+  def document_has_republishable_editions?
+    (latest_unpublished_edition || live_edition || pre_publication_edition).present?
   end
 
-  def the_document_has_been_unpublished?
-    latest_unpublished_edition.present?
+  def latest_unpublished_edition
+    return @latest_unpublished_edition if defined?(@latest_unpublished_edition)
+
+    @latest_unpublished_edition = document.editions.unpublished.last
```
[the remark this answers is not in the document]
Comment (ryanb-gds) [reply app/workers/publishing_api_document_republishing_worker.rb:54]: Following discussion we agreed to leave these in place, as having more specific getters for `published_edition` and `withdrawn_edition` is considered to help making the consuming code more explicit

**Review by ryanb-gds — APPROVED** [review]
Thanks for trying to make some improvements here and for generally investing in making Whitehall a better place! As discussed on the call, might be worth adding some coverage for the bulk publication flag to the [integration test](https://github.com/alphagov/whitehall/blob/main/test/integration/workers/publishing_api_document_republishing_worker_integration_test.rb), but feel free to do that in a separate PR, and only if you think it's worth it.

I have a suspicion that the next developers to run into this code will probably feel like they still need to refactor it again according to their preferences before they can make sense of it, just because it's so fundamentally complicated, but let's see. If this helps you deliver the republishing UI work for now then I see no reason not to merge it.

File: app/workers/publishing_api_document_republishing_worker.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (yndajas) [reply app/workers/publishing_api_document_republishing_worker.rb:54]: I think additionally we wanted to reduce the length of the public method, and this was a thing that could easily be hidden away as an implementation detail. But as discussed, adding the `published_edition` and `withdrawn_edition` makes things more explicit in a few places

File: test/unit/app/workers/publishing_api_document_republishing_worker_test.rb [hunk 2]
```diff
@@ -4,104 +4,467 @@
 class PublishingApiDocumentRepublishingWorkerTest < ActiveSupport::TestCase
```
[the remark this answers is not in the document]
Comment (ryanb-gds) [reply test/unit/app/workers/publishing_api_document_republishing_worker_test.rb:4]: Thanks, that's good to know

**Pull request merged**